pre_commit_linter.py

```diff
@@ -0,0 +1,311 @@
+        python pre_commit_linter.py
+
+2.  To lint all files in the folder or to lint just a specific file
+        python pre_commit_linter.py --path filepath
+
+3.  To lint a specific list of files (*.py only). Separate files by spaces
+        python pre_commit_linter.py --files file_1 file_2 ... file_n
+
+Note that the root folder MUST be named 'oppia-ml-proto'.
+"""
+
+# Pylint has issues with the import order of argparse.
+# pylint: disable=wrong-import-order
+import argparse
+import fnmatch
+import multiprocessing
+import os
+import subprocess
+import sys
+import time
+# pylint: enable=wrong-import-order
+
+_PARSER = argparse.ArgumentParser()
+_EXCLUSIVE_GROUP = _PARSER.add_mutually_exclusive_group()
+_EXCLUSIVE_GROUP.add_argument(
+    '--path',
+    help='path to the directory with files to be linted',
+    action='store')
+_EXCLUSIVE_GROUP.add_argument(
+    '--files',
+    nargs='+',
+    help='specific files to be linted. Space separated list',
+    action='store')
+
+if not os.getcwd().endswith('oppia-ml-proto'):
+    print('')
+    print('ERROR    Please run this script from the oppia root directory.')
+
+_PROTOTOOL_PATH = os.path.join(
+    os.getcwd(), 'third_party', 'prototool-1.10.0', 'prototool')
+
+if not os.path.exists(_PROTOTOOL_PATH):
+    print('')
+    print('ERROR    Please run install_prototool.sh first to install prototool')
+    sys.exit(1)
+
+_MESSAGE_TYPE_SUCCESS = 'SUCCESS'
+_MESSAGE_TYPE_FAILED = 'FAILED'
+
+EXCLUDED_PATHS = ('third_party/*',)
+
+PROTOTOOL_CONFIG_FILE = 'prototool_config.json'
+
+def _get_changed_filenames():
+    """Returns a list of modified files (both staged and unstaged)
+
+    Returns:
+        a list of filenames of modified files.
+    """
+    unstaged_files = subprocess.check_output([
+        'git', 'diff', '--name-only',
+        '--diff-filter=ACM']).splitlines()
+    staged_files = subprocess.check_output([
+        'git', 'diff', '--cached', '--name-only',
+        '--diff-filter=ACM']).splitlines()
+    return unstaged_files + staged_files
+
+
+def _get_all_files_in_directory(dir_path):
+    """Recursively collects all files in directory and
+    subdirectories of specified path.
+
+    Args:
+        dir_path: str. Path to the folder to be linted.
+
+    Returns:
+        a list of files in directory and subdirectories without excluded files.
+    """
+    files_in_directory = []
+    for _dir, _, files in os.walk(dir_path):
+        for file_name in files:
+            filename = os.path.relpath(
+                os.path.join(_dir, file_name), os.getcwd())
+
+            files_in_directory.append(filename)
+    return files_in_directory
+
+
+def _lint_proto_files(files_to_lint, result, config):
+    """Prints a list of lint errors in the given list of Proto files.
+
+    Args:
+        files_to_lint: list(str). A list of filepaths to lint.
+        result: multiprocessing.Queue. A queue to put results of test.
+    """
+    start_time = time.time()
```

Comment:
Perhaps, conceptually, this should be moved lower down, after you've checked whether there are files to lint.

Reply:
Done